template/utils/sync.py

```diff
@@ -0,0 +1,234 @@
+# The MIT License (MIT)
+# Copyright © 2023 Yuma Rao
+# Copyright © 2023 Opentensor Foundation
+
+# Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+# documentation files (the “Software”), to deal in the Software without restriction, including without limitation
+# the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software,
+# and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+
+# The above copyright notice and this permission notice shall be included in all copies or substantial portions of
+# the Software.
+
+# THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO
+# THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+# DEALINGS IN THE SOFTWARE.
+
+# Utils for checkpointing and saving the model.
+import copy
+import torch
+import bittensor as bt
+from typing import List
+
+import template
+
+spec_version = template.__spec_version__
+
+
+def check_registered(self):
+    """
+    Checks if the miner or validator hotkey is registered on the network.
+    """
+    if self.wallet.hotkey.ss58_address not in self.metagraph.hotkeys:
+        bt.logging.error(
+            f"\nYour validator: {self.wallet} if not registered to chain connection: {self.subtensor} \nRun btcli register and try again."
+        )
+        exit()
+
+
+def sync(self):
```

Comment:
I think I am more in favour of these functions being in the template base miner/validator. This will expose them to the users where it will be easier for subnet creators to alter them as needed. For instance, subnet owners may want to know what happens to their weights before it it submited to the chain. Currently the design puts these functions quite deep into the code which makes it hard for users to interact with.